### PR TITLE
Context: adding ability to add custom load callback

### PR DESCRIPTION
**Pull request by steweg**
This patch adds ability to add custom load callback and also disable searchdirs options

### CODE REVIEW

File: libyang/context.py [hunk 1]
```diff
@@ -19,25 +19,71 @@
...
+        "_module_data_clb",
+        "_cffi_handle",
+        "_cdata_modules",
+        "__dict__",
+    )
 
     def __init__(
         self,
         search_path: Optional[str] = None,
+        disable_searchdirs: bool = False,
```
Comment (samuel-gauthier) [inline libyang/context.py:68]: Could you put that new feature in a separate commit?

File: tests/test_context.py [hunk 2]
```diff
@@ -111,3 +111,18 @@ def test_ctx_leafref_extended(self):
         with Context(YANG_DIR, leafref_extended=True) as ctx:
             mod = ctx.load_module("yolo-leafref-extended")
             self.assertIsInstance(mod, Module)
+
+    def test_ctx_disable_searchdirs_and_clb(self):
+        def get_module_clb(mod_name, *_):
+            YOLO_NODETYPES_MOD_PATH = os.path.join(YANG_DIR, "yolo/yolo-nodetypes.yang")
+            self.assertEqual(mod_name, "yolo-nodetypes")
```
Comment (samuel-gauthier) [inline tests/test_context.py:118]: Let's have two tests, those features are not that related.

File: libyang/context.py [hunk 3]
```diff
@@ -19,25 +19,71 @@
...
+    free_module_data[0] = lib.lypy_module_imp_data_free_clb
+    instance = ffi.from_handle(user_data)
+    in_fmt, content = instance.get_module_data(
+        c2str(mod_name), c2str(mod_rev), c2str(submod_name), c2str(submod_rev)
+    )
+    if content is None:
+        return lib.LY_ENOT
+    fmt[0] = schema_in_format(in_fmt)
+    module_data[0] = content
+    return lib.LY_SUCCESS
```
Comment (samuel-gauthier) [inline libyang/context.py:52]: Just to understand, what are you using this for in the end?

Comment (steweg) [reply libyang/context.py:52]: My application is downloading the YANG files from the remote server. To do that you need to use this kind of callback.

File: libyang/context.py (hunk 1 above)
Comment (steweg) [reply libyang/context.py:68]: Separated into #104, the rest of the code is kept in this PR

File: tests/test_context.py (hunk 2 above)
Comment (steweg) [reply tests/test_context.py]: Ok, done within #104

**Comment by steweg** [comment]
On the testing level this is dependent on #104, so please take look on #104 first. thanks

File: libyang/context.py (hunk 1 above)
Comment (rjarry) [reply libyang/context.py:68]: There was no need to create a separate pull request. This makes this one harder to test. Can you push the two separate commits in this PR and close the other?

Thanks

**Review by rjarry — CHANGES_REQUESTED** [review]
Can you please put the commit from #104 in this pull request? It will simplify testing.

Thanks.

**Comment by rjarry** [comment]
Hi,

libyang-python only supports the master branch of C libyang (currently 2.x).

The lydevel check is only meant for informative purposes.

https://github.com/CESNET/libyang-python/blob/f09ed1164d860fd9f04c008e42bbac13ccdb2ff2/tox.ini#L18-L20

If it fails (currently it should fail because libyang 3.x is under development) it is not a problem. 

Could you remove all references to libyang 3 which was not released yet and only focus on the enhancements that you require?

Thank you.

**Comment by steweg** [comment]
> Hi,
> 
> libyang-python only supports the master branch of C libyang (currently 2.x).
> 
> The lydevel check is only meant for informative purposes.
> 
> https://github.com/CESNET/libyang-python/blob/f09ed1164d860fd9f04c008e42bbac13ccdb2ff2/tox.ini#L18-L20
> 
> If it fails (currently it should fail because libyang 3.x is under development) it is not a problem.
> 
> Could you remove all references to libyang 3 which was not released yet and only focus on the enhancements that you require?
> 
> Thank you.

Done

**Comment by samuel-gauthier** [comment]
Sorry for the delay on this... I pushed the first commit, but I don't understand what the second one does... Could you explain in more detail what you use it for, what it does exactly, and how it works?

**Comment by steweg** [comment]
It allows you to define your own callback functions to load YANG modules if they are not found on in standard path. This is considered as sort of last resort option, which allows you to store YANG modules in remote/central locations and download it using some HTTP(s) client etc... libyang just tells you what file & version you should get, and it is up to your callback implementation to try to get it.

File: libyang/context.py [hunk 4]
```diff
@@ -19,9 +19,48 @@
...
+@ffi.def_extern(name="lypy_module_imp_data_free_clb")
+def libyang_c_module_imp_data_free_clb(cdata, user_data):
+    instance = ffi.from_handle(user_data)
+    instance.free_module_data(cdata)
+
+
+# -------------------------------------------------------------------------------------
+@ffi.def_extern(name="lypy_module_imp_clb")
+def libyang_c_module_imp_clb(
+    mod_name,
```
Comment (samuel-gauthier) [inline libyang/context.py:32]: Please add docstring and describe the arguments.

File: libyang/context.py [hunk 5]
```diff
@@ -468,3 +511,41 @@ def __iter__(self) -> Iterator[Module]:
         while mod:
             yield Module(self, mod)
             mod = lib.ly_ctx_get_module_iter(self.cdata, idx)
+
+    def free_module_data(self, cdata) -> None:
+        self._cdata_modules.remove(cdata)
+
+    def get_module_data(
+        self,
```
Comment (samuel-gauthier) [inline libyang/context.py:519]: Please add docstring and describe the arguments.

File: libyang/context.py [hunk 6]
```diff
@@ -468,3 +511,41 @@ def __iter__(self) -> Iterator[Module]:
...
+            mod_name, mod_rev, submod_name, submod_rev
+        )
+        if module_data is None:
+            return fmt_str, None
+        module_data_c = str2c(module_data)
+        self._cdata_modules.append(module_data_c)
+        return fmt_str, module_data_c
+
+    def set_module_data_clb(
+        self,
```
Comment (samuel-gauthier) [inline libyang/context.py:537]: Please add docstring and describe the arguments.

File: libyang/context.py [hunk 7]
```diff
@@ -468,3 +511,41 @@ def __iter__(self) -> Iterator[Module]:
         while mod:
             yield Module(self, mod)
             mod = lib.ly_ctx_get_module_iter(self.cdata, idx)
+
+    def free_module_data(self, cdata) -> None:
+        self._cdata_modules.remove(cdata)
```
Comment (samuel-gauthier) [inline libyang/context.py:516]: Please add docstring and describe the arguments, especially what is in cdata here?

File: libyang/context.py [hunk 8]
```diff
@@ -468,3 +511,41 @@ def __iter__(self) -> Iterator[Module]:
...
+    ) -> Tuple[str, Optional[str]]:
+        if self._module_data_clb is None:
+            return None
+        fmt_str, module_data = self._module_data_clb(
+            mod_name, mod_rev, submod_name, submod_rev
+        )
+        if module_data is None:
+            return fmt_str, None
+        module_data_c = str2c(module_data)
+        self._cdata_modules.append(module_data_c)
```
Comment (samuel-gauthier) [inline libyang/context.py:533]: Why do you need to store this?

File: libyang/context.py [hunk 9]
```diff
@@ -19,9 +19,48 @@
...
+    submod_rev,
+    user_data,
+    fmt,
+    module_data,
+    free_module_data,
+):
+    fmt[0] = lib.LYS_IN_UNKNOWN
+    module_data[0] = ffi.NULL
+    free_module_data[0] = lib.lypy_module_imp_data_free_clb
+    instance = ffi.from_handle(user_data)
```
Comment (samuel-gauthier) [inline libyang/context.py:44]: How about putting only the function here? I don't like giving access to the whole context object here, it seems overkill.

File: libyang/context.py (hunk 4 above)
Comment (steweg) [reply libyang/context.py:32]: Done

File: libyang/context.py (hunk 5 above)
Comment (steweg) [reply libyang/context.py:519]: Done, but the example is pretty pointless, as this API is called by libyang directly and not by user. Check the unit test that shows real example how to use this

File: libyang/context.py (hunk 6 above)
Comment (steweg) [reply libyang/context.py:537]: Done

File: libyang/context.py (hunk 7 above)
Comment (steweg) [reply libyang/context.py:516]: Done

File: libyang/context.py (hunk 8 above)
Comment (steweg) [reply libyang/context.py:533]: The data provided are python string. That one needs to held as libyang is using those data directly. It doesn't create internal copy. Therefore the data must stay in memory until libyang will ask to free them. If it would not be held by context itself, it will be up to user to hold them and release them when not needed anymore

File: libyang/context.py (hunk 9 above)
Comment (steweg) [reply libyang/context.py:44]: Done

**Comment by samuel-gauthier** [comment]
Pushed with minor changes, thanks!

**Comment by samuel-gauthier** [comment]
@steweg there is a problem with this pull request, can you look at this ignored exception? It can be reproduced by launching tox.

pypy3: commands[0] /home/runner/work/libyang-python/libyang-python/tests> python -Wd -m unittest discover -c
................Exception ignored from cffi callback <function libyang_c_module_imp_clb at 0x00007f523c4bfc40>:
Traceback (most recent call last):
  File "/home/runner/work/libyang-python/libyang-python/.tox/pypy3/lib/pypy3.9/site-packages/libyang/context.py", line 71, in libyang_c_module_imp_clb
    ret = instance.get_module_data(
  File "/home/runner/work/libyang-python/libyang-python/.tox/pypy3/lib/pypy3.9/site-packages/libyang/context.py", line 140, in get_module_data
    fmt_str, module_data = self._module_data_clb(
TypeError: cannot unpack non-iterable NoneType object
........................................................................................................................
----------------------------------------------------------------------
Ran 136 tests in 0.502s

Thanks!

File: libyang/context.py [hunk 10]
```diff
@@ -133,12 +133,11 @@ def get_module_data(
             Tuple of format string and YANG (sub)module schema
         """
         if self._module_data_clb is None:
-            return "", None
-        fmt_str, module_data = self._module_data_clb(
-            mod_name, mod_rev, submod_name, submod_rev
-        )
-        if module_data is None:
-            return fmt_str, None
+            return None
```
Comment (samuel-gauthier) [inline libyang/context.py:136]: It made sense to return "", None, do you really want to change it?

Comment (steweg) [reply libyang/context.py:136]: According to `get_module_data() -> Optional[Tuple[str, str]]:`  it shall return either None or Tuple of strings. So it was a invalid return value.
